articles/112_generated_interfaces_cpp.md

```diff
@@ -0,0 +1,127 @@
+---
+layout: default
+title: Generated C++ interfaces
+permalink: articles/generated_interfaces_cpp.html
+abstract:
+  This article describes the generated C++ code for ROS 2 interfaces.
+author: '[Dirk Thomas](https://github.com/dirk-thomas)'
+---
+
+* This will become a table of contents (this text will be scraped).
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+
+## Scope
+
+This article specifies the generated C++ code for ROS interface types defined in the [interface definition article](interface_definition.html).
+
+
+## Namespacing
+
+All code of a ROS package should be defined in a namespace named after the package.
+To separate the generated code from other code within the package it is defined in a sub namespace:
+* namespace for ROS messages: `package_name::msg`.
+* namespace for ROS services: `package_name::srv`.
+
+
+## Generated files
+
+Following the C++ style guide of ROS 2 the namespace hierarchy is mapped to a folder structure.
+The filenames use lowercase alphanumeric characters with underscores for separating words and end with either `.hpp` or `.cpp`.
+
+<div class="alert alert-warning" markdown="1">
+  <b>TODO:</b> describe separate header files
+</div>
+
+
+### Messages
+
+A message results in a templated `struct` with the same name followed by an underscore.
+The single template argument is the allocator for the data structure.
+
+For ease of use there is a `typedef` with the same name as the message which uses a default allocator (e.g. `std::allocator`).
```

Comment:
I've always wondered about this. Why don't we just have a default template argument like with `std::vector`?

http://en.cppreference.com/w/cpp/container/vector

Is it because it's the only template argument?